app/models/audience_segment.rb

```diff
@@ -0,0 +1,54 @@
+class AudienceSegment < ApplicationRecord
+  # enum does not like names that start with "not_"
+  enum type_of: {
+    testing: 0, # never matches anyone, used in test factory
+    trusted: 1,
+    posted: 2,
+    no_posts_yet: 3,
+    dark_theme: 4,
+    light_theme: 5,
+    no_experience: 6,
+    experience1: 7,
+    experience2: 8,
+    experience3: 9,
+    experience4: 10,
+    experience5: 11
+  }
+
+  has_many :segmented_users, dependent: :destroy
+  has_many :users, through: :segmented_users
+
+  after_validation :run_query
+
+  QUERIES = {
+    testing: ->(scope = User) { scope.where(id: nil) },
+    trusted: ->(scope = User) { scope.with_role(:trusted) },
+    no_posts_yet: ->(scope = User) { scope.where(articles_count: 0) },
+    posted: ->(scope = User) { scope.where("articles_count > 0") },
+    dark_theme: ->(scope = User) { scope.where(id: Users::Setting.dark_theme.select(:user_id)) },
+    light_theme: ->(scope = User) { scope.where(id: Users::Setting.light_theme.select(:user_id)) },
+    experience1: ->(scope = User) { scope.with_experience_level(1) },
+    experience2: ->(scope = User) { scope.with_experience_level(2) },
+    experience3: ->(scope = User) { scope.with_experience_level(3) },
+    experience4: ->(scope = User) { scope.with_experience_level(4) },
+    experience5: ->(scope = User) { scope.with_experience_level(5) },
+    no_experience: ->(scope = User) { scope.with_experience_level(nil) }
+  }.freeze
+
+  def self.all_users_in_segment(symbol, scope: User)
+    query_for_segment(symbol)&.call(scope) || []
+  end
+
+  def self.query_for_segment(symbol)
+    QUERIES[symbol.to_sym]
+  end
+
+  def run_query
+    self.users = self.class.all_users_in_segment(type_of)
+  end
+
+  def refresh!
```

Comment:
I imagine we refresh the segments on a user-by-user basis after they've taken key actions?

Reply:
@benhalpern My naïve assumption would be a daily cron that refreshes all segments, but probably scope `User` to our "recently active" (I don't think this is a scope, yet, but something we've done more-or-less [manually](https://github.com/forem/forem/blob/408bb14283bd90773fbc3ed74cdc06ef8ddc8c39/app/services/notifications/notifiable_action/send.rb#L40) so far).

One easy enhancement might be only refresh segments actively used by approved & published billboards, and have a backgrounded callback from billboards to refresh a "stale" associated segment on save.

We can also do something to refresh segments on user activity (your suggestion here), but it'll be trickier, and would probably not be my first step. Some reasons it's tricky:

* We can refresh any segment the user is _in_, but that won't add the user to any segments where they are missing but _should now_ be in. The net effect is that we'd need the user-triggered refresh to walk _all_ the segments to add or remove the user id as required (though some will presumably be no-ops).
* Ensuring that we cover all the "key actions" for any/all segments is going to be a bit of a moving target initially — with what we have now, we're looking at role changes, article publication, and settings update, but I expect this will evolve rapidly at first, before eventually settling down. Even defining something like "article publication" can be a bit of a rabbit-hole.

All that said, I imagine we'll get there, but I'd prefer to be more confident that the basic implementation is viable before we over-optimize. Plus, with ad-triggered refresh _and_ a fast-enough cron refresh, we _might_ be able to increase the cron frequency _instead_ of fine-tuning the "key actions" refresh.

Reply:
Just circling up here. This all makes sense and is well thought out.